### PR TITLE
Don't emit clutz aliases for reexported symbols.

### PR DESCRIPTION
**Pull request by LucasSloan**
Tsickle should only emit clutz aliases in the original exporting file, not any files that reexport.  Reexporting a symbol doesn't bring the symbol into the file, so the alias would refer to a symbol that isn't available.

### CODE REVIEW

**Comment by mprobst** [comment]
@LucasSloan is this PR no longer necessary? What happened? :-)

**Comment by LucasSloan** [comment]
Whoops, I meant to close the other PR, since it didn't pass tests and needed a major rethink.  Please review!

File: src/tsickle.ts [hunk 1]
```diff
@@ -2130,6 +2130,23 @@ export function emitWithTsickle(
   };
 }
 
+function isSymbolDeclaredInSourceFile(symbol: ts.Symbol, sourceFile: ts.SourceFile) {
+  const declarations = symbol.getDeclarations();
+  if (!declarations) {
+    // If there are no declarations, err on the side of safety and emit the clutz
+    // aliases
```
Comment (evmar) [inline src/tsickle.ts]: This comment is a bit weird because it comes out of nowhere (it looks like the function is just checking where a symbol lives, but now it's talking about clutz).  Maybe a better comment is a function doc that says "if there are no declarations, returns true" without talking about clutz and then documenting the clutz thing at the usage site?  Not sure.

File: src/tsickle.ts [hunk 2]
```diff
@@ -2130,6 +2130,23 @@ export function emitWithTsickle(
...
+  for (const decl of declarations) {
+    if (decl.getSourceFile() === sourceFile) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
 function addClutzAliases(
```
Comment (evmar) [inline src/tsickle.ts]: I have no idea what this function is supposed to do, heh.

File: src/tsickle.ts (hunk 1 above)
Comment (evmar) [reply src/tsickle.ts]: Another idea: make this function take the declaration list rather than the symbol itself, and have the caller pre-filter out symbols that have no declaration.

File: test_files/reexport.declaration/reexport.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export * from './export';
```
Comment (evmar) [inline test_files/reexport.declaration/reexport.ts]: Can you add some comments to these test files about what they're testing, like what property of the output is important?

File: test/test_support.ts [hunk 4]
```diff
@@ -120,6 +120,12 @@ export function createSourceCachingHost(
     if (sources.has(fileName)) {
       return true;
     }
+    // Typescript occasionally needs to look on disk for files we don't pass into
+    // the program as a source (eg to resolve a module that's in node_modules),
+    // but only .ts files explicitly passed in should be findable
+    if (/.*\.ts$/.test(fileName)) {
+      return false;
+    }
```
Comment (evmar) [inline test/test_support.ts]: Why is this necessary in this change?

File: src/tsickle.ts (hunk 1 above)
Comment (LucasSloan) [reply src/tsickle.ts]: Good suggestion, that was bugging me!

File: src/tsickle.ts (hunk 2 above)
Comment (LucasSloan) [reply src/tsickle.ts]: Added a comment - does it make sense?

File: test/test_support.ts (hunk 4 above)
Comment (LucasSloan) [reply test/test_support.ts]: Test support provides a kind of hermetic way to run a typescript compilation, where you explicitly pass in the sources as strings.  However, we have one test that needs to check if a file exists on disk to do module resolution inside node_modules (even though it doesn't actually read the contents of the file).

I added a test case where I wanted to compile 2 .d.ts's to check them for consistency, but without this line, tsc would notice that there was a .ts file around and try to grab that instead of the .d.ts file I wanted to compile.

This was the easiest way I could find to resolve the 2 use cases.

File: test_files/reexport.declaration/reexport.ts (hunk 3 above)
Comment (LucasSloan) [reply test_files/reexport.declaration/reexport.ts]: Added a comment.

File: test/test_support.ts [hunk 5]
```diff
@@ -120,6 +120,12 @@ export function createSourceCachingHost(
     if (sources.has(fileName)) {
       return true;
     }
+    // Typescript occasionally needs to look on disk for files we don't pass into
+    // the program as a source (eg to resolve a module that's in node_modules),
+    // but only .ts files explicitly passed in should be findable
+    if (/.*\.ts$/.test(fileName)) {
```
Comment (evmar) [inline test/test_support.ts]: Can write it:
  /\.ts$/

Comment (LucasSloan) [reply test/test_support.ts]: Great, thanks for the tip!
